schemachanger: add ALTER TABLE ... ALTER COLUMN ... SET DEFAULT

annrpom opened the pull request:
This PR implements `ALTER COLUMN SET DEFAULT` in the declarative
schema changer.

Epic: CRDB-31283

Fixes: #116247
Release note: None

annrpom committed:
schemachanger: add ALTER TABLE ... ALTER COLUMN ... SET DEFAULT

This PR implements `ALTER COLUMN SET DEFAULT` in the declarative
schema changer.

Epic: CRDB-31283

Fixes: #116697
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116697)
<!-- Reviewable:end -->

annrpom commented:
> update: as of v23.2, alter table b alter column x set default 'hi'; is unimplemented in the declarative schema changer (DSC) and will still leave dangling back-reference in the type descriptor. But, starting from v22.1, we supported DROP TYPE in declarative schema changer, and drop type a; , under declarative schema changer, will be able to proceed to drop this type despite it having an dangling back-reference to a table that's already dropped. The legacy schema changer, on the other hand, is not "smart" enough to recognize that and will block the dropping.

tried testing this - confirming we are able to successfully drop the type

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/scgraph/graph.go:
```diff
@@ -110,6 +110,9 @@ func New(cs scpb.CurrentState) (*Graph, error) {
 			Attrs:    []rel.Attr{screl.ReferencedFunctionIDs},
 			Inverted: true,
 		},
+		{
+			Attrs: []rel.Attr{screl.Expr},
+		},
```
I'm not very familiar with this part. Could you explain why this is needed?

annrpom replied:
Ah - thank you for catching this; just bits left from testing. Excluded from my PR

Xiang-Gu commented on pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_set_default/alter_table_alter_column_set_default.definition:
```diff
@@ -0,0 +1,22 @@
+setup
+CREATE TABLE t (i INT PRIMARY KEY, j INT);
+ALTER TABLE t ALTER COLUMN j SET DEFAULT 2;
```
nit: since this is setup, we could probably just do one line `create table t (i int primary key, j int default 2);`

annrpom replied:
done

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_default.go:
```diff
@@ -0,0 +1,42 @@
+) {
+	alterColumnPreChecks(b, tn, tbl, t.Column)
+	colID := getColumnIDFromColumnName(b, tbl.TableID, t.Column, true /* required */)
+	// Block alters on system columns.
+	scpb.ForEachColumn(
+		b.QueryByID(tbl.TableID),
+		func(_ scpb.Status, _ scpb.TargetStatus, e *scpb.Column) {
+			if e.ColumnID == colID {
+				// Block drops on system columns.
+				panicIfSystemColumn(e, t.Column.String())
+			}
+		})
```
nit: I think scpb.ForEachColumn is deprecated. Let's use something like 

```
panicIfSystemColumn(mustRetrieveColumnElem(b, tbl.TableID, colID), t.Column.String())
```

and then kindly update `mustRetrieveColumnElem` to use the now-preferred way of retrieving a column element
```
func mustRetrieveColumnElem(
	b BuildCtx, tableID catid.DescID, columnID catid.ColumnID,
) (column *scpb.Column) {
	return b.QueryByID(tableID).FilterColumn().Filter(func(
		current scpb.Status, target scpb.TargetStatus, e *scpb.Column,
	) bool {
		return e.ColumnID == columnID
	}).MustGetOneElement()
}
```

annrpom replied:
done

annrpom replied:
I changed this in `ALTER ... SET NOT NULL` as well (first commit)

Xiang-Gu reviewed (commented):
thanks for doing this. I left some early comments and questions. The overall approach looks good to me. Let's now focus on getting the CI failures fixed.

annrpom committed:
schemachanger: refactor deprecated call in ALTER ... SET NOT NULL

This patch refactors the deprecated `ForEachColumn` logic.

Epic: none

Release note: None

annrpom commented:
@Xiang-Gu Want to take another look 👀 ?

annrpom commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_default.go:
```diff
@@ -0,0 +1,145 @@
+
+		// Check if this reference is cross DB.
+		if tblNamespace.DatabaseID != seqNamespace.DatabaseID {
+			if err := b.CanCreateCrossDBSequenceOwnerRef(); err != nil {
```
Ah! Ignore this comment - got lost in all the diff cross ref cluster setttings

fqazi reviewed (approved):
Great work!  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 14 of 14 files at r2, 128 of 128 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116697#-:-NrGYrXY1n8EONzlkT0i:bp67qck)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scdeps/sctestdeps/test_state.go` line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116697#-NrGXWJh5idHV6cnjXdj:-NrGXWJiE6mRGlpfP3Fi:bv3jn9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1341ae063905cfcf78a491d7a23c6478a22319dd/pkg/sql/schemachanger/scdeps/sctestdeps/test_state.go#L224)):*
> ```Go
> // CanCreateCrossDBSequenceRef implements scbuild.SchemaFeatureCheck.
> func (s *TestState) CanCreateCrossDBSequenceRef() error {
> 	return nil
> ```

Should we log a side effect for this?


<!-- Sent from Reviewable.io -->

annrpom commented on pkg/sql/schema_changer.go:
```diff
@@ -3316,3 +3316,17 @@ func (p *planner) CanCreateCrossDBSequenceOwnerRef() error {
+
+// CanCreateCrossDBSequenceRef returns if cross database sequence
+// references are allowed.
+func (p *planner) CanCreateCrossDBSequenceRef() error {
```
> Should we log a side effect for this?

That's for testing(TestState)  - the actual change ^ here will block the user from doing this

annrpom commented:
TFTR! ('-')7

bors r=fqazi
